src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/AbstractDocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,232 @@
+                        title,
+                        c => GetProjectFixesAsync(fixAllContext.WithCancellationToken(c), fixAllContext.Project),
+                        nameof(DocumentBasedFixAllProvider)));
+
+                case FixAllScope.Solution:
+                    return Task.FromResult<CodeAction?>(CodeAction.Create(
+                        title,
+                        c => GetSolutionFixesAsync(fixAllContext.WithCancellationToken(c)),
+                        nameof(DocumentBasedFixAllProvider)));
+
+                case FixAllScope.Custom:
+                default:
+                    return Task.FromResult<CodeAction?>(null);
+            }
+        }
+
+#pragma warning restore RS0005 // Do not use generic 'CodeAction.Create' to create 'CodeAction'
+
+        private async Task<Document> GetDocumentFixesAsync(FixAllContext fixAllContext)
+        {
+            RoslynDebug.AssertNotNull(fixAllContext.Document);
+
+            var document = fixAllContext.Document;
+            var documentDiagnosticsToFix = await FixAllContextHelper.GetDocumentDiagnosticsToFixAsync(fixAllContext).ConfigureAwait(false);
+            if (!documentDiagnosticsToFix.TryGetValue(document, out var diagnostics))
+                return document;
+
+            var newDoc = await this.FixAllAsync(fixAllContext, document, diagnostics).ConfigureAwait(false);
+            return newDoc ?? document;
+        }
+
+        private Task<Solution> GetProjectFixesAsync(FixAllContext fixAllContext, Project project)
+            => FixAllInSolutionAsync(fixAllContext, ImmutableArray.Create(project.Id));
+
+        private Task<Solution> GetSolutionFixesAsync(FixAllContext fixAllContext)
+            => FixAllInSolutionAsync(fixAllContext);
+
+        private Task<Solution> FixAllInSolutionAsync(FixAllContext fixAllContext)
+        {
+            var solution = fixAllContext.Solution;
+            var dependencyGraph = solution.GetProjectDependencyGraph();
+
+            // Walk through each project in topological order, determining and applying the diagnostics for each
+            // project.  We do this in topological order so that the compilations for successive projects are readily
+            // available as we just computed them for dependent projects.  If we were to do it out of order, we might
+            // start with a project that has a ton of dependencies, and we'd spend an inordinate amount of time just
+            // building the compilations for it before we could proceed.
+            //
+            // By processing one project at a time, we can also let go of a project once done with it, allowing us to
+            // reclaim lots of the memory so we don't overload the system while processing a large solution.
+            var sortedProjectIds = dependencyGraph.GetTopologicallySortedProjects().ToImmutableArray();
+            return FixAllInSolutionAsync(fixAllContext, sortedProjectIds);
```

Comment:
the topological sort is a key part to make consistent and even progress while doing fix-all.  it ensures we don't get to a project, and then have to spend an inordinate time right at hte beginning producing all the dependent compilations/references for a project.